### PR TITLE
PR for #22: deploy shiny

### PR DESCRIPTION
**Pull request by MosesStewart**
In this issue we . . .

- Move the shiny code from https://github.com/JMSLab/BootstrapReport/pull/21#issue-2022175617 into the repository in a new directory called `/app`
    - Update default behavior to use first column in data file uploaded
- Add text to point users to `BootstrapReport` repository to bottom of interface.
- Deployed working version of website at [https://mosesstewart02.shinyapps.io/bootstrapreport/](https://mosesstewart02.shinyapps.io/bootstrapreport/)

Adding @jmshapir

### CODE REVIEW

File: shiny/app.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    @output
+    @render.table
+    def stats() -> object:
+        ooi = eval_ooi()
+        ooi.get_sk_min()
+        df_stats = pd.DataFrame(columns = ['SK distance', 'Minimum SK distance', 'SK minimizing mean', 'SK minimizing SD'])
+        df_stats.at[0, 'SK distance'] = ooi.sk_dist
+        df_stats.at[0, 'Minimum SK distance'] = ooi.skmin
+        df_stats.at[0, 'SK minimizing mean'] = ooi.skmin_mean
+        df_stats.at[0, 'SK minimizing SD'] = ooi.skmin_sd
```
Comment (jmshapir) [inline shiny/app.py:109]: I put in the values here:
https://github.com/JMSLab/BootstrapReport/blob/3fb6719ed5d8a090c161c263c63bdf739ec83152/examples/_generate_example_data.py#L43
and used the [`normal_replicates`](https://github.com/JMSLab/BootstrapReport/blob/3fb6719ed5d8a090c161c263c63bdf739ec83152/examples/normal_replicates.csv) as the example data.

I was expecting to get a low SK since I used the mean and SD that generated the data. Instead the SK is pretty large and the p-p plot is very bowed. 

Am I missing something or is it possible there's a bug somewhere in one of the pathways in the code?

image

File: shiny/app.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+from BootstrapReport import ObjectOfInterest
+
+app_ui = ui.page_fixed(
+    ui.div(style = "height:30px"),
+    ui.panel_title("Analyze a Report"),
+
+    ui.layout_sidebar(
+
+      ui.panel_sidebar(
+        ui.input_numeric("est", "Estimate", 0),
```
Comment (jmshapir) [inline shiny/app.py:15]:
```suggestion
        ui.input_numeric("est", "Point estimate", 0),
```

**Review by jmshapir — CHANGES_REQUESTED** [review]
@MosesStewart thanks! Very excited to see this. 

I had a few notes and will now have a look at steps to deploy to my shinyapps instance.

**Comment by jmshapir** [comment]
@MosesStewart I sent you via separate methods a [token and secret](https://docs.posit.co/shinyapps.io/getting-started.html#configure-rsconnect) that can be used to deploy via the `jmslab` shiny instance.

After addressing the points above please deploy to `jmslab` as `BootstrapReport`.

**Comment by jmshapir** [comment]
@rcalvo12 I'm adding you as reviewer here. It would be ideal if you can review after @MosesStewart has addressed the batch of comments above and deployed to the `jmslab` instance.

Thanks!

File: shiny/app.py (hunk 1 above)
Comment (MosesStewart) [reply shiny/app.py:109]: I do not have a great answer for you, besides that I added extra randomness in the data generating process by generating the variables from a $\mathcal{N}(\mu, n \times \sigma)$ distribution:

https://github.com/JMSLab/BootstrapReport/blob/a9710406b7c8720e564e7e3eea5f3c57ef9a019e/examples/_generate_example_data.py#L44-L45

. . . and performing a bootstrap below:

https://github.com/JMSLab/BootstrapReport/blob/a9710406b7c8720e564e7e3eea5f3c57ef9a019e/examples/_generate_example_data.py#L69-L73

So we're relying on (1), a sample of (100) random variables to well-approximate a normal distribution with large variance and (2), a draw of (100) replicates from those random variables to well-approximate the distribution of those (100) random variables.

If we use the mean and sample-std we get a much nicer looking plot:
image

File: shiny/app.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                                "Column Count": [column_count],
+                                "Column Names": [column_names],})
+        return info_df
+
+    @output
+    @render.plot(alt="pp plot", width = 670, height = 670)
+    def pp_plot() -> object:
+        ooi = eval_ooi()
+        fig = ooi.pp_plot(fontsize = 13.5, legend_fontsize = 14.5, labelsize = 18)
+        return fig
```
[the remark this answers is not in the document]
Comment (MosesStewart) [reply shiny/app.py]: Updated in b2e686c !

File: shiny/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+matplotlib >= 3.8.0
+scipy >= 1.11.1
+numpy >= 1.26.0
+seaborn >= 0.13.0
+pandas >= 2.1.3
+xlrd >= 2.0.0
+pyreadr >= 0.5.0
+shiny >= 0.6.0
+Jinja2 >= 3.1.2
+git+https://github.com/JMSLab/BootstrapReport
```
[the remark this answers is not in the document]
Comment (MosesStewart) [reply shiny/requirements.txt]: Updated in b2e686c !

File: shiny/app.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+app_ui = ui.page_fixed(
+    ui.div(style = "height:30px"),
+    ui.panel_title("Analyze a Report"),
+
+    ui.layout_sidebar(
+
+      ui.panel_sidebar(
+        ui.input_numeric("est", "Estimate", 0),
+        ui.input_numeric("se", "Standard error", 1),
+        ui.input_file("repcsv", "Choose data file", accept=[".csv", ".mat", ".dta", ".rds", ".xls", ".xlsx"], multiple=False),
```
[the remark this answers is not in the document]
Comment (MosesStewart) [reply shiny/app.py]: Added above the box because spacing was a little odd in b2e686c. Let me know if that's okay!

File: shiny/requirements.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (MosesStewart) [reply shiny/requirements.txt:10]: Had to undo in 45cafb2. Since we're installing from github, pip can't interpret the version numbers.

File: README.md [hunk 6]
```diff
@@ -8,5 +8,8 @@ Documentation:
 * [Methods](./BootstrapReport.lyx)
 * [Examples](./examples/)
 
+WebApp:
```
[the remark this answers is not in the document]
Comment (jmshapir) [reply README.md:11]: > @jmshapir I think maybe you already deployed a package called `bootstrapreport`, because after I deployed it got auto-renamed to `bootstrapreport1`. I don't have access to delete deployed packages to `jmslab` in [shinyapps](https://www.shinyapps.io/admin/#/applications/all), so if we want to re-deploy I think we need to delete both and redo it.

@MosesStewart there was an undeployed app called `bootstrapreport` which I've now archived and deleted. That should (?) clear the namespace for a new deployment. For now I haven't removed `bootstrapreport1` so that it remains available for testing.

File: examples/_generate_example_data.py [hunk 7]
```diff
@@ -37,16 +37,15 @@ def generate_example_dataset(name):
...
     if name == 'normal':
         num_rep, mean, std = 100, 0.74, 0.086
-        raw_data = norm.rvs(size = num_rep, loc = mean, scale = np.sqrt(num_rep) * std)
+        raw_data = rng.normal(size = num_rep, loc = mean, scale = np.sqrt(num_rep) * std)
         replicates = get_replicates(raw_data)
         estimate = np.median(raw_data)
     elif name == 'gamma':
         num_rep, variance, mean = 499, 0.186, 2.189
-        raw_data = gamma.rvs((mean**2)/(variance * num_rep), scale = (variance * num_rep)/mean, size = 50)
+        raw_data = rng.gamma((mean**2)/(variance * num_rep), scale = mean/(variance * num_rep), size = 50)
```
Comment (jmshapir) [inline examples/_generate_example_data.py:48]: @MosesStewart was this change intentional? It has a big effect on the values in the example.

File: shiny/app.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+app_ui = ui.page_fixed(
+    ui.div(style = "height:30px"),
+    ui.panel_title("BootstrapReport"),
+
+    ui.layout_sidebar(
+
+      ui.panel_sidebar(
+        ui.input_numeric("est", "Point estimate", 0),
+        ui.input_numeric("se", "Standard error", 1),
+        ui.input_file("repcsv", "Choose data file (CSV, MAT, DTA, RDS, RDATA, XLS, XLSX)", accept = [".csv", ".mat", ".dta", ".rds", ".rdata", ".xls", ".xlsx"], multiple = False),
```
Comment (jmshapir) [inline shiny/app.py:17]: Hit an error with XLSX below.

@MosesStewart have you tested with this file format?

image

**Review by jmshapir — CHANGES_REQUESTED** [review]
@MosesStewart thanks! I took another pass through and flagged a few more points.

@rcalvo12 I think it's fine for you to review anytime now. As part of your review, in addition to checking code, it would be great if you can try out a bunch of different inputs (e.g., file formats, column names, etc.) to the [app](https://jmslab.shinyapps.io/bootstrapreport1/) to make sure it works the way you expect.

File: examples/_generate_example_data.py (hunk 7 above)
Comment (MosesStewart) [reply examples/_generate_example_data.py:48]: [`np.random.seed`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html) is depreciated, so I updated the code to use `np.random.default_rng()`.

Comment (jmshapir) [reply examples/_generate_example_data.py:48]: @MosesStewart thanks! If you look at the `git diff`, you'll see a big change in the replicate values. I'm guessing that means we implicitly changed the type of gamma we're simulating. I'm guessing that was unintentional?

File: shiny/app.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        info_df = pd.DataFrame({"Row Count": [row_count],
+                                "Column Names": [column_names],})
+        return info_df
+
+    @output
+    @render.plot(alt="pp plot", width = 670, height = 670)
+    def pp_plot() -> object:
+        ooi = eval_ooi()
+        if ooi == None:
+            raise AttributeError("Plot will appear after replicates are uploaded")
```
[the remark this answers is not in the document]
Comment (MosesStewart) [reply shiny/app.py]: (I changed the default behavior, because you cannot change it so it doesn't appear to be an error)

File: shiny/app.py
[the remark this answers is not in the document]
Comment (rcalvo12) [reply shiny/app.py:1]: @MosesStewart On the table, would it not be feasible to just export the info_df as a csv?

[the remark this answers is not in the document]
Comment (MosesStewart) [reply shiny/app.py:1]: We could do that

File: examples/gamma_estimate.csv [hunk 10]
```diff
@@ -1,2 +1,2 @@
 estimate,std_err
-3.329153904463066,2.0256286444979983
+1.8450286322049927,0.3898786195682134
```
Comment (jmshapir) [inline examples/gamma_estimate.csv:2]: Pretty meaningful change here. Is this just luck of the (random) draw?

File: shiny/README.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+# Shiny Implementation
```
Comment (jmshapir) [inline shiny/README.md:1]: Let's add provenance for the `ratio` files to the readme by including a stable link to their source.

File: shiny/app.py
[the remark this answers is not in the document]
Comment (jmshapir) [reply shiny/app.py:1]: > @jmshapir, do you see value in adding a button that would export the csv of the table and png of the plot?

@rcalvo12 @MosesStewart I do but since the plot can be downloaded by right-clicking and the table is only four numbers I think this is functionality we can set aside for now.

**Review by jmshapir — CHANGES_REQUESTED** [review]

File: examples/gamma_estimate.csv (hunk 10 above)
Comment (jmshapir) [reply examples/gamma_estimate.csv:2]: OK! It's anyway just a hypothetical example. Just want to have some confidence we know how we're generating it.

File: shiny/README.md (hunk 11 above)
Comment (MosesStewart) [reply shiny/README.md:1]: Addressed in 8246918 !

**Review by rcalvo12 — APPROVED** [review]

**Comment by jmshapir** [comment]
> I accidentally made another url at https://jmslab.shinyapps.io/bootstrapreport1/ at first, but then successfully updated the one at https://jmslab.shinyapps.io/bootstrapreport/

@MosesStewart no problem! I removed the extra instance.

Going forward if we want to test changes that we are not ready to deploy to the main instance, we can create a duplicate called `bootstrapreportdev`.

**Comment by jmshapir** [comment]
@MosesStewart it [looks like](https://github.com/JMSLab/BootstrapReport/actions/runs/7081229675/job/19270208973) one test job got canceled? Might just be because we merged before it was done. In that case, no problem, we can run it next time we need to change something. Thanks!
